libpromises/evalfunction.c

```diff
@@ -1165,17 +1185,24 @@ static FnCallResult FnCallVariablesMatching(EvalContext *ctx, ARG_UNUSED const P
 
     {
         VariableTableIterator *iter = EvalContextVariableTableIteratorNew(ctx, NULL, NULL, NULL);
-        StringSet *global_matches = VariablesMatching(ctx, iter, finalargs);
+        JsonElement *global_matches = VariablesMatching(ctx, fp, iter, finalargs, fulldata);
+        VariableTableIteratorDestroy(iter);
 
-        StringSetIterator it = StringSetIteratorInit(global_matches);
-        const char *element = NULL;
-        while ((element = StringSetIteratorNext(&it)))
+        assert (JsonGetContainerType(global_matches) == JSON_CONTAINER_TYPE_OBJECT);
+
+        if (fulldata)
         {
-            RlistPrepend(&matches, element, RVAL_TYPE_SCALAR);
+            return (FnCallResult) { FNCALL_SUCCESS, (Rval) { global_matches, RVAL_TYPE_CONTAINER } };
         }
 
-        StringSetDestroy(global_matches);
-        VariableTableIteratorDestroy(iter);
+        JsonIterator jiter = JsonIteratorInit(global_matches);
+        const char *key;
+        while ((key = JsonIteratorNextKey(&jiter)))
+        {
+            RlistPrepend(&matches, key, RVAL_TYPE_SCALAR);
```

Comment:
What are the odds of `key` not being a string? If zero, then `assert()`. If it's possible, then print a message and skip.